### PR TITLE
Start from library method

### PR DESCRIPTION
**Pull request by annakocheshkova**
Implemented `startFromLibrary` for iOS and Android. 
NOTE: need to give an ability to turn off autostart from native library.

### CODE REVIEW

File: Assets/AppCenter/AppCenterBehavior.cs [hunk 1]
```diff
@@ -8,7 +8,6 @@
 using System.Reflection;
 using Microsoft.AppCenter.Unity.Internal;
 using System.Linq;
-
```
Comment (MatkovIvan) [inline Assets/AppCenter/AppCenterBehavior.cs]: Revert empty line

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Shared/StartupType.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        /// <summary>
+        /// Analytics service is started from library.
+        /// </summary>
+        OneCollector = 3,
+
+        /// <summary>
+        /// Nothing is configured on startup.
+        /// </summary>
+        None = 4
+    }
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Shared/StartupType.cs]: Please add the missing types

File: Assets/AppCenter/AppCenterBehavior.cs [hunk 3]
```diff
@@ -77,7 +76,10 @@ private void StartAppCenter()
         var nativeServiceTypes = AppCenter.ServicesToNativeTypes(services);
         AppCenterInternal.Start(appSecret, nativeServiceTypes, services.Length);
 #endif
-
+        if (settings.startupType == StartupType.OneCollector) {
+            var analyticsService = new Type[1]{ AppCenterSettings.Analytics };
+            AppCenterInternal.StartFromLibrary(AppCenter.ServicesToNativeTypes(analyticsService));
```
Comment (MatkovIvan) [inline Assets/AppCenter/AppCenterBehavior.cs]: Move it to the transmission tab

File: Assets/AppCenter/AppCenterBehavior.cs [hunk 4]
```diff
@@ -77,7 +76,10 @@ private void StartAppCenter()
         var nativeServiceTypes = AppCenter.ServicesToNativeTypes(services);
         AppCenterInternal.Start(appSecret, nativeServiceTypes, services.Length);
 #endif
-
+        if (settings.startupType == StartupType.OneCollector) {
+            var analyticsService = new Type[1]{ AppCenterSettings.Analytics };
```
Comment (MatkovIvan) [inline Assets/AppCenter/AppCenterBehavior.cs]: `new Type[1]` -> `new []`

File: Assets/AppCenter/AppCenterBehavior.cs [hunk 5]
```diff
@@ -77,7 +76,10 @@ private void StartAppCenter()
         var nativeServiceTypes = AppCenter.ServicesToNativeTypes(services);
         AppCenterInternal.Start(appSecret, nativeServiceTypes, services.Length);
 #endif
-
+        if (settings.startupType == StartupType.OneCollector) {
```
Comment (MatkovIvan) [inline Assets/AppCenter/AppCenterBehavior.cs]: Why is this condition necessary?

File: Assets/AppCenter/AppCenterSettings.cs [hunk 6]
```diff
@@ -37,6 +37,8 @@ public class AppCenterSettings : ScriptableObject
     [Tooltip("To enable push for Android apps, you need to set the Sender ID found in the Firebase portal for your application.")]
     public string SenderId;
 
+    public StartupType startupType = StartupType.AppCenter;
```
Comment (MatkovIvan) [inline Assets/AppCenter/AppCenterSettings.cs]: ~Please add `Tooltip` for editor~

Do you sure that it should be visible for ALL AppCenter users in the UI?

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 7]
```diff
@@ -8,6 +8,12 @@
 
 namespace Microsoft.AppCenter.Unity.Internal
 {
+#if UNITY_IOS || UNITY_ANDROID
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: Why is `UNITY_IOS` here (in the `/Android` folder)?

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 8]
```diff
@@ -84,6 +90,25 @@ private static AndroidJavaObject GetAndroidApplication()
...
+        {
+            AndroidJavaClass unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
+            AndroidJavaObject activity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity");
+            return activity.Call<AndroidJavaObject>("getApplicationContext");
+        }
+
+        public static void StartFromLibrary(ServiceType[] servicesArray)
+        {    
+            IntPtr services = servicesArray[0];
+            IntPtr appCenterRawClass = AndroidJNI.FindClass("com/microsoft/appcenter/AppCenter");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: There is cached one in the `_appCenter`

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 9]
```diff
@@ -84,6 +90,25 @@ private static AndroidJavaObject GetAndroidApplication()
             wrapperSdkObject.Call("setLiveUpdatePackageHash", liveUpdatePackageHash);
             _appCenter.CallStatic("setWrapperSdk", wrapperSdkObject);
         }
+        
+        private static AndroidJavaObject GetAndroidContext()
+        {
+            AndroidJavaClass unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: Let's cache it

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 10]
```diff
@@ -84,6 +90,25 @@ private static AndroidJavaObject GetAndroidApplication()
...
+            AndroidJavaObject activity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity");
+            return activity.Call<AndroidJavaObject>("getApplicationContext");
+        }
+
+        public static void StartFromLibrary(ServiceType[] servicesArray)
+        {    
+            IntPtr services = servicesArray[0];
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: Let's use `var` where possible (this is rule for  C# from our .NET SDK).

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs [hunk 11]
```diff
@@ -9,6 +9,12 @@
 
 namespace Microsoft.AppCenter.Unity.Internal
 {
+#if UNITY_IOS || UNITY_ANDROID
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs]: This is folder for UWP

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs [hunk 12]
```diff
@@ -145,6 +151,11 @@ private static Microsoft.AppCenter.LogLevel LogLevelFromUnity(int logLevel)
             }
         }
 
+
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs]: Extra line

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/iOS/AppCenterInternal.cs [hunk 13]
```diff
@@ -8,6 +8,12 @@
 
 namespace Microsoft.AppCenter.Unity.Internal
 {
+#if UNITY_IOS || UNITY_ANDROID
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/iOS/AppCenterInternal.cs]: This is folder for iOS only

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.h [hunk 14]
```diff
@@ -11,6 +11,7 @@ extern "C" void appcenter_unity_set_log_url(const char* logUrl);
 extern "C" void appcenter_unity_set_enabled(bool isEnabled);
 extern "C" bool appcenter_unity_is_enabled();
 extern "C" const char* appcenter_unity_get_install_id();
+extern "C" void appcenter_unity_start_from_library(__strong Class classes[], int count);
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.h]: Why not pointer?

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm [hunk 15]
```diff
@@ -32,6 +32,15 @@ void appcenter_unity_set_enabled(bool isEnabled)
   [MSAppCenter setEnabled:isEnabled];
 }
 
+void appcenter_unity_start_from_library(__strong Class classes[], int count) {
+    NSMutableArray<Class>* mutableClasses = [[NSMutableArray alloc] init];
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm]: `[NSMutableArray new]`?

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm [hunk 16]
```diff
@@ -32,6 +32,15 @@ void appcenter_unity_set_enabled(bool isEnabled)
...
+    for (int i=0; i<count; i++) {
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm]: Add missing spaces

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm [hunk 17]
```diff
@@ -32,6 +32,15 @@ void appcenter_unity_set_enabled(bool isEnabled)
...
+        [mutableClasses addObject:classes[i]];
+    }
+
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.mm]: Extra line

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.h (hunk 14 above)
Comment (annakocheshkova) [reply Assets/AppCenter/Plugins/iOS/Core/AppCenterUnity.h]: It works only in this state

**Comment by annakocheshkova** [comment]
Decided to implement startupMode independently from this.

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 18]
```diff
@@ -84,6 +85,28 @@ private static AndroidJavaObject GetAndroidApplication()
             wrapperSdkObject.Call("setLiveUpdatePackageHash", liveUpdatePackageHash);
             _appCenter.CallStatic("setWrapperSdk", wrapperSdkObject);
         }
+        
+        private static AndroidJavaObject GetAndroidContext()
+        {
+            if (_context != null) {
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: Move `{` to next line

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 19]
```diff
@@ -84,6 +85,28 @@ private static AndroidJavaObject GetAndroidApplication()
...
+                return _context;
+            }
+            AndroidJavaClass unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: `var`

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 20]
```diff
@@ -84,6 +85,28 @@ private static AndroidJavaObject GetAndroidApplication()
...
+            AndroidJavaObject activity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: `var`

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs [hunk 21]
```diff
@@ -84,6 +85,28 @@ private static AndroidJavaObject GetAndroidApplication()
...
+            _context = activity.Call<AndroidJavaObject>("getApplicationContext");
+            return _context;
+        }
+
+        public static void StartFromLibrary(IntPtr[] servicesArray)
+        {    
+            IntPtr services = servicesArray[0];
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Android/AppCenterInternal.cs]: I think we can get rid of this variable

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs [hunk 22]
```diff
@@ -145,6 +145,10 @@ private static Microsoft.AppCenter.LogLevel LogLevelFromUnity(int logLevel)
             }
         }
 
+        public static void StartFromLibrary(Type[] services) {
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs]: Please format

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/iOS/AppCenterInternal.cs [hunk 23]
```diff
@@ -68,6 +68,10 @@ public static void SetCustomProperties(IntPtr properties)
                                                 liveUpdatePackageHash);
         }
 
+        public static void StartFromLibrary(IntPtr[] services) {
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/iOS/AppCenterInternal.cs]: Please format

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/Blank/AppCenterInternal.cs [hunk 24]
```diff
@@ -33,6 +33,12 @@ public static void StartServices(ServiceType[] services, int numServices)
         {
         }
 
+
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/Blank/AppCenterInternal.cs]: Extra line

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**